### ISSUE TITLE
BUGS with For Loop nodes

### ISSUE THREAD

**Issue by mifth** [nortikin/sverchok#4267]
## Problem statement

I have found **2 issues** with the For Loops.

**1 Issue:**
Using Range Mode creates less Total Loops.
![image](https://user-images.githubusercontent.com/2251298/128556756-dae34a7f-629e-4ded-a5fe-0716ccd47899.png)

**2 Issue:**
In Range Mode there is no Data 0 socket. I cannot add anything to the output.
![image](https://user-images.githubusercontent.com/2251298/128556971-a91df12f-fe78-4bde-bf77-eb5cd7760481.png)

Here Is my blend file. I used Master branch.
[tmp1.zip](https://github.com/nortikin/sverchok/files/6947403/tmp1.zip)

**Comment by zeffii** [comment]
try a DebugPrint node instead of Stethoscope, it will print the incoming data to the system console.

`Stethoschope` won't display the concatenation of all received inputs of every iteration of the loop, rather it should theoretically be displaying all input but it changes too fast to see it, it looks like displays only the last input it receives... 
   - so if the loop takes 5 ms, and you have 5 different values going into Stethoscope, then each loop-iteration would be visible for only 1ms, and the last iteration would be permanently visible because there is no followup update.

`DebugPrint` on the other hand will record all input into the system console, and you see what it received (at a human readable time frame) on each iteration.

... i think.

**Comment by mifth** [comment]
Wait. Total Loops is always a constant value. It must be always the same.

**Comment by zeffii** [comment]
i don't see the point of reading that value then..

**Comment by zeffii** [comment]
.. things should not break by reading it, but it seems not useful?

**Comment by mifth** [comment]
This is a useful value. I don't need to add additional nodes to calculate it. I used it in the same way in Houdini. We just need to fix it.

**Comment by zeffii** [comment]
i'm sure you're right,  i'm not at a computer now : /

**Comment by mifth** [comment]
no worries

**Comment by zeffii** [comment]
did you look at the examples? https://github.com/nortikin/sverchok/blob/master/docs/nodes/logic/loop_in.rst

**Comment by mifth** [comment]
I saw it.

**Comment by Durman** [comment]
1. There is property to limit maximum number of loops. It is 5 by default and locates on property panel.
2. `Loop out` node will only get socket if correspondent socket of `Loop in` node is connected.

**Comment by mifth** [comment]
This all sounds strange.
1. total loops should show the length of a list. It makes no sense to show loops per frame. Or please rename the total loops to loops per frame.
2. I see no sense in such a limitation. Why it's not possible to add a socket to the loop out node? Basically, all modes do the same.

I see only confusion in it.

**Comment by Durman** [comment]
1. Why per frame? I think the total loops output is `min(iterations, max_iterations)`
2. It is possible. You just have to connect `Loop in` node first. There is no sence to create sockets for `Loop out` node separately because they work together with Loop in sockets.
![2021-08-07_10-54-40](https://user-images.githubusercontent.com/28003269/128591523-5c6339a6-951f-4f97-b4f6-378cdb99780f.png)

**Comment by mifth** [comment]
Thanks for the reply. It looks I had to use the Number Range node instead of the List Length node. But even if I use it then the Total Loops value = 0! 
It must be = 5!

![image](https://user-images.githubusercontent.com/2251298/128610830-ffb6ca0e-94bf-41ba-badb-78cc359a1973.png)

**Comment by mifth** [comment]
I have attached my blend file about my previous post. 
[tmp1.zip](https://github.com/nortikin/sverchok/files/6949582/tmp1.zip)

**Comment by Durman** [comment]
I really caught one bug, created separate issue #4269 for it.
My previous example is really bad. It does not show how loop node works. It would be better to use `add` operator in the math node.
![loop example](https://user-images.githubusercontent.com/28003269/128624839-46b43572-a33c-4b6a-879a-2cfefc2accfd.gif)
Now the work should be more clear.
I don't think that the iterations socket can gen multiple values. In your example it just takes first value (`0`) and use for generating number of loops.

**Comment by mifth** [comment]
Then there is a bug. I assume Total Values parameter should show all values of an iteration like it does  in the For Each mode.

**Comment by Durman** [comment]
I'm not quite understand what do you mean.

**Comment by mifth** [comment]
Let me write it in python. Here is an example of using TotalLoops. I suppose it works in this way:

```
someList = [1, 2, 30, 1, 0]
totalLoops = 5

for i in range(5):
   # just an example of totalLoops usage
    if someList[i] > totalLoops:
        break
```
    
Total Loops value should be always equal to length of a range. Is it correct?

**Comment by vicdoval** [comment]
> Thanks for the reply. It looks I had to use the Number Range node instead of the List Length node. But even if I use it then the Total Loops value = 0!
> It must be = 5!
> 
> ![image](https://user-images.githubusercontent.com/2251298/128610830-ffb6ca0e-94bf-41ba-badb-78cc359a1973.png)

The iteration  socket will get only  the first value so it is 0 iterations

Maybe this brings some light
![image](https://user-images.githubusercontent.com/10011941/128701018-f4897491-70cb-48bf-87b1-2502f7f50f1f.png)

**Comment by Durman** [comment]
@mifth I can't commit your code but I can write code of this example https://github.com/nortikin/sverchok/issues/4267#issuecomment-894757849
```py
data = data_socket.read_input_data()  # [[1, 2, 3, 4]]
iterations = min(iterations_socket.read_input_data()[0][0], max_iterations_property)  # 5

for i in range(iterations):
    data = math_add(data, 1)
    loop_number = i
    loop_total = iterations

assert(data, [[6, 7, 8, 9]])
```

**Comment by mifth** [comment]
Thanks to all for explanation. I see that the Loop Total is what should be. 

One thing. Why there is limitation of loops? I mean the **max_iterations_property**. If I will parse a mesh with 1000000 of points and I would like to use the Range mode? Or maybe 10mln of points? This limitation has no sense.

**Comment by mifth** [comment]
> @mifth I can't commit your code but I can write code of this example [#4267 (comment)](https://github.com/nortikin/sverchok/issues/4267#issuecomment-894757849)
> 
> ```python
> data = data_socket.read_input_data()  # [[1, 2, 3, 4]]
> iterations = min(iterations_socket.read_input_data()[0][0], max_iterations_property)  # 5
> 
> for i in range(iterations):
>     data = math_add(data, 1)
>     loop_number = i
>     loop_total = iterations
> 
> assert(data, [[6, 7, 8, 9]])
> ```

Btw, an input of the Iterations parameter should be a number. Not an array.
![image](https://user-images.githubusercontent.com/2251298/128728931-d81154f9-15a1-4fc3-bf8f-dd7280bf889e.png)
![image](https://user-images.githubusercontent.com/2251298/128729159-73fc854e-b52e-4171-8fb5-cb816168a546.png)

**Comment by mifth** [comment]
I solved All issues which are on my first post. I consider them fixed in my mind. The only confusion is LIMITATION. Can we remove it  or switch it off. Modern computers can pass millions of iterations these days.

**Comment by zeffii** [comment]
The justification for `max_iterations` is that there is a potential for the loop to produce a ridiculous amount of geometry. It's a remnant from a previous (now deprecated) looping mechanism. In particular this would be noticeable if the loop operates on all incoming geometry and produces a multiple of the original elements - the definition of exponential growth. In that scenario accidentally setting the number of iterations from 5 to 6 could take a minute of processing (for example), and from 5 to 7 could take hours or never finish.
- This was an issue for the older update system
- This was primarily an issue for certain kinds of looping scenarios (exponential growth)

We could certainly add the option to toggle the hard `max_iterations` limit off, there are scenarios where the loop will not generate exponential information growth. Maybe because nodetree processing can be cancelled now it's time to make the `max_iterations` optional.

**Comment by Durman** [comment]
When / if the work of the loop nodes will be cancelable there will be no sence in having the `max_iterations` property.

**Comment by mifth** [comment]
@zeffii yeah, but Esc key should cancel any iterations like in houdini. But there is no such a feateure. And I believe even this feature does not exist then anyway there is no sense in this limitation. Devs must be always understand what they do. I believe that a modern node system should not have such a strange limitation from 90s.

@Durman  Is it hard to implement the Cancel System? I'm just curious about it. I don't remember id animation nodes have one?

**Comment by Durman** [comment]
Current update system already can be cancelled between work of two nodes. Loop nodes take control over a tree execution and current update system can't stop their work. To make loop nodes stoppable their code should be fixed. Now it should not be too hard. Node groups for example are stoppable.

**Comment by mifth** [comment]
Is this mean that we can expect that the limitation will be removed in future?

**Comment by Durman** [comment]
Yes, but there are also a lot of other things which would be nice to have. For example using multithreading for executing separate tree parts in parallel, using Cython for boosting performance, make work of nodes stoppable, support of muting nodes etc.

**Comment by mifth** [comment]
Thanks.
